### DISCUSSION TITLE
Get numeric values of Decimal data types defined in schema.

### DISCUSSION THREAD

**Discussion by lucasdavidferrero** [prisma/prisma#23968] [Q&A]
### Question

When I run ` prisma.aikon_articulo.findMany()` I get as a String the values of fields that were defined as Decimal. I know that under the hood Prisma is using decimal.js library. 
I'd like to get them as numeric values and not strings, just like the database stores it. Take the example of 'aik_ar_cosnet' field: SQL Stores the numeric value 109457.00, but Prisma library returns "109457" string. I'd like that Prisma returns the same value the database does.

Aside for that, I'm defining Decimal for those values that are used for currency. I don't know if this Prisma data type is correct to use in this situation. Maybe I could use another Prisma datatype?

### Expected behavior (optional)

Get numeric values from fields that were defined as Decimal in Prisma.

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
model aikon_articulo {
  aik_ar_codigo           String @db.VarChar(24) @id
  aik_ar_descri           String? @db.VarChar(200)
  aik_ar_memo             String? @db.VarChar(4000)
  aik_re1_codigo          String? @db.VarChar(5)
  aik_re2_codigo          String? @db.VarChar(5)
  aik_ar_alto             Float?
  aik_ar_ancho            Float?
  aik_ar_profundo         Float?
  aik_ar_color            String? @db.VarChar(20)
  aik_ar_peso             Float?
  aik_ar_descria          String? @db.VarChar(200)
  aik_ar_fechamodif       DateTime?
  aik_ar_mesesgarantia    Float?
  aik_ar_cosnet           Decimal? @db.Decimal(15,2)
  aik_ap_utilidad         Decimal? @db.Decimal(6, 2)
  aik_impuesto_interno    Decimal? @db.Decimal(6, 2)
  aik_iva_porcen          Decimal? @db.Decimal(6, 2)
  aik_stock_total         Int? @db.UnsignedSmallInt
  aik_ap_precio_iva       Decimal? @db.Decimal(15,2)
}
```

```ts
prisma.aikon_articulo.findMany()
```

- OS: Windows
- Database: MySQL
- Node.js version: v18.14.1

- Prisma version: 5.9.1

**Comment by ludralph** [comment]
Hi @lucasdavidferrero 👋 

Prisma uses the `Decimal.js` library to handle Decimal fields. This is because JavaScript's native number type can lead to precision errors, especially when dealing with currency or other high-precision numbers. Therefore, Decimal is indeed a suitable choice for currency fields in your Prisma schema.
However, when you fetch data from the database, Prisma returns Decimal values as instances of Decimal.js, not as native JavaScript numbers. This is why you're seeing the values as strings when you try to print them out or serialize them to JSON.
If you want to convert a Decimal.js instance to a JavaScript number, you can use the `toNumber` method. Here's an example:

```js
const data = await prisma.aikon_articulo.findMany();
data.forEach(item => {
  item.aik_ar_cosnet = item.aik_ar_cosnet.toNumber();
});
```

**Reply by ElhanM** [reply]
Ideally I would like to avoid having to do this every time I fetch

There should be an option we can use in order to tell prisma to always fetch them as numbers

**Reply by lucasdavidferrero** [reply]
Also, we are forced to define numbers as Decimal in data structures if we use Typescript.

**Reply by AnirudhMemani** [reply]
What is the type of Decimal?

**Reply by ElhanM** [reply]
It is an object under the hood, but it is not that difficult to work with when you just stick to the Decimal type and not convert it to a number everywhere

**Reply by CarlosTrapet** [reply]
@ElhanM 
> There should be an option we can use in order to tell prisma to always fetch them as numbers

Like @ludralph mentioned, you would lose precision by converting it to numbers directly. Perhaps in your use case using another type of numeral type would be more appropriate?
If you 100% need them as numbers, you could create a DTO layer above your Prisma fetching layer that translates them `toNumber()`

But then again, if that needs to be exposed via an HTTP call to a different service, the result would be inaccurate and your initial service directly calling Prisma would be making implementation decisions for other services that consume it

**Reply by CarlosTrapet** [reply]
@lucasdavidferrero 
> so, we are forced to define numbers as Decimal in data structures if we use Typescript.

what do you mean? You are not forced to use any specific types; you could use numbers or integers or floats where it makes sense in your use case, and use Decimals where you want point precision. 
You can then export the type to share with consumers:
`export type PrismaDecimal = Prisma.Decimal`

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
